torch/utils/data/dataset.py

```diff
@@ -35,3 +35,51 @@ def __getitem__(self, index):
 
     def __len__(self):
         return self.data_tensor.size(0)
+
+
+class ConcatDataset(Dataset):
+    """
+    Dataset to concatenate multiple datasets.
+    Purpose: useful to assemble different existing datasets, possibly
+    large-scale datasets as the concatenation operation is done in an
+    on-the-fly manner.
+    Args:
+        datasets (iterable): List of datasets to be concatenated
+    """
+
+    @staticmethod
+    def cumsum(sequence):
+        result = []
+        sum = 0
+        for s in sequence:
+            l = len(s)
+            cumsum.append(l + sum)
+            sum += l
+        return result
+
+    @staticmethod
+    def searchsorted(idx, sequence):
+        for i in range(1, len(sequence)):
+            if sequence[i - 1] < idx <= sequence[i]:
+                return i
+
+    def __init__(self, datasets):
+        super(ConcatDataset, self).__init__()
+
+        self.datasets = list(datasets)
+        assert len(datasets) > 0, 'datasets should not be an empty iterable'
+        self.cum_sizes = self.cumsum(self.datasets)
+
+    def __len__(self):
+        return self.cum_sizes[-1]
+
+    def __getitem__(self, idx):
+        super(ConcatDataset, self).__getitem__(idx)
+        dataset_index = self.searchsorted(idx, self.cum_sizes)
```

Comment:
I think you can just use `bisect` from Python which does binary search:

```
dataset_index = bisect.bisect_left(self.cum_sizes, idx)
```

Raise an `IndexError` if index is out-of-bounds